scripts: west_commands: extend `build.dir-fmt` format args

thorsten-klein opened the pull request:
`build.dir-fmt` format string arguments are extended. 

New supported format args are:
- `{west_topdir}` which is the absolute path to the west workspace directory (west topdir). This is helpful for users when they want to place all build folders under a fix path within the west workspace.

- `{source_dir_workspace}` which is the relative path of the source directory to the west workspace (west topdir). If the source directory is outside of the west workspace, `{source_dir_workspace}` will be relative to the filesystem root. `{source_dir_workspace}` is helpful to keep builds in separate folders under a specific subfolder. Since the path is relative to west topdir , it does not matter in which working directory the `west build` command was called. 

Background: `{source_dir}` cannot be used currently since it is relative to the current working directory and it might therefore be a relative path starting with `../`. `{source_dir_ws}` will never start with `../`.

E.g. users can now use following `dir-fmt` in order to store builds under a fix path `<west_workspace>/build/`:

```
dir-fmt = {west_topdir}/build/{board}/{source_dir_workspace}
```

thorsten-klein committed:
scripts: west_commands: extend build.dir-fmt format args

build.dir-fmt format string arguments are extended. New format args are
'west_topdir' (absolute path to the west workspace directory) and
'source_dir_workspace' (relative path of the source directory to
west_topdir).

Signed-off-by: Thorsten Klein <thorsten.klein@bshg.com>

marc-hb reviewed (commented):
Looking at all the `os.cwd()` (which should probably be `Path.cwd()`), I wonder about people using `west` outside the west workspace, thanks to the ZEPHYR_BASE variable. How is this going to work for them? Has `dir-fmt` ever worked for them? It's OK if it does not but then it should fail clearly and "gracefully".

Noob questions sorry.

thorsten-klein commented on scripts/west_commands/tests/test_build.py:
```diff
@@ -55,3 +59,59 @@ def test_parse_remainder(test_case):
+
+
+cwd = Path(os.getcwd())
+west_ws_dir = cwd.parent
```
Reworked.

tejlmand commented on doc/develop/west/build-flash-debug.rst:
```diff
@@ -393,12 +393,19 @@ You can :ref:`configure <west-config-cmd>` ``west build`` using these options.
+           current working directory. If the current working directory is
+           inside the source directory, this is an empty string. If no source
+           directory is specified, it defaults to current working directory.
+         - ``source_dir_ws``: Path to the CMake source directory relative to
```
`ws` is not a common abreviation, and as this is not something users has to type daily then I would prefer to have it spelled out, like this:
```suggestion
         - ``source_dir_workspace``: Path to the source directory relative to
```

tejlmand commented on scripts/west_commands/build.py:
```diff
@@ -456,16 +456,34 @@ def _update_cache(self):
+    def _get_dir_fmt_context(self):
+        # Return a dictionary of build attributes which are used while
+        # substituting the placeholders in the build.dir-fmt format string.
+        west_top_dir = west_topdir()
```
```suggestion
        west_topdir = west_topdir()
```

It's called `topdir` everywhere else, so why the extra `_` in `top_dir` ?
https://docs.zephyrproject.org/latest/develop/west/basics.html

thorsten-klein replied:
I wanted to avoid that the imported west_topdir is shadowed. Is this common?

marc-hb commented on doc/develop/west/build-flash-debug.rst:
```diff
@@ -395,10 +395,17 @@ You can :ref:`configure <west-config-cmd>` ``west build`` using these options.
+           current working directory. If the current working directory is
+           inside the source directory, this is an empty string. If no source
+           directory is specified, it defaults to current working directory.
+         - ``source_dir_workspace``: Path to the source directory, relative to
```
My bad, I went too fast and didn't realize the previous is relative too... I think examples would be great here.

thorsten-klein commented:
Is there anything else I should provide or address to get this feature merged? 💯

thorsten-klein committed:
scripts: west_commands: added tests for config build.dir-fmt

Tests are added for different build.dir-fmt format string arguments.

Signed-off-by: Thorsten Klein <thorsten.klein@bshg.com>

thorsten-klein committed:
doc: develop: west: config option build.dir-fmt

new format arguments for config option build.dir-fmt are documented
accordingly.

Signed-off-by: Thorsten Klein <thorsten.klein@bshg.com>

thorsten-klein commented on scripts/west_commands/build.py:
```diff
@@ -287,11 +287,11 @@ def _find_board(self):
         elif 'BOARD' in os.environ:
             board, origin = os.environ['BOARD'], 'env'
-        elif self.config_board is not None:
+        elif getattr(self, 'config_board', None) and self.config_board is not None:
```
This statement failed since config_board might not be set.
In tests `_get_dir_fmt_context` is called directly (and not `do_run` where the variable is assigned).
So an exception is raised when `self.config_board` is tried to be accessed.

thorsten-klein replied:
Thank you. I was blind.

thorsten-klein commented:
Ref: https://docs.zephyrproject.org/latest/contribute/style/python.html#excluded-files

> It is encouraged for contributors, when changing an excluded file, to remove it from the list and format it in a separate commit.

Do you want me to do this for `build.py` in this PR?
